lib/pos_staking_rewards.go

```diff
@@ -0,0 +1,282 @@
+			// and return early here.
+			return 0, 0, errors.Errorf("computeStakerRewardAndValidatorCommission: validator entry should never be nil")
+		}
+
+		if validatorEntry.DelegatedStakeCommissionBasisPoints > 0 {
+			// We use integer math to compute the validator's commission. The commission is computed as:
+			// floor(stakerReward * validatorCommissionBasisPoints / 10000)
+			validatorCommissionNanos = computeValidatorCommission(
+				stakerRewardNanos, validatorEntry.DelegatedStakeCommissionBasisPoints,
+			)
+
+			if validatorCommissionNanos.Cmp(stakerRewardNanos) > 0 {
+				// This should never happen. If the validator's commission is greater than the total staker reward amount,
+				// then something has gone wrong.
+				return 0, 0, errors.Errorf(
+					"computeStakerRewardAndValidatorCommission: validator commission is greater than staker reward amount",
+				)
+			}
+
+			// Deduct the validator commission from the staker's reward.
+			stakerRewardNanos.Sub(stakerRewardNanos, validatorCommissionNanos)
+		}
+	}
+
+	// At this point, we have the staker's reward and the validator's commission. We need to convert them
+	// to uint64s and return them.
+	if !stakerRewardNanos.IsUint64() || !validatorCommissionNanos.IsUint64() {
+		return 0, 0, errors.Errorf(
+			"computeStakerRewardAndValidatorCommission: staker reward or validator commission is not a uint64",
+		)
+	}
+
+	return stakerRewardNanos.Uint64(), validatorCommissionNanos.Uint64(), nil
+}
+
+func (bav *UtxoView) distributeStakingReward(validatorPKID *PKID, stakerPKID *PKID, rewardNanos uint64) error {
+	// Fetch the staker's latest StakeEntry.
+	stakeEntry, err := bav.GetStakeEntry(validatorPKID, stakerPKID)
+	if err != nil {
+		return errors.Wrapf(err, "distributeStakingReward: problem fetching staker's StakeEntry: ")
+	}
+
+	// At this point, there are three possible cases:
+	// 1. The stake entry still exists and wants to restake their rewards.
+	// 2. The stake entry still exists and does not want to restake their rewards.
+	// 3. The stake entry has unstaked since the snapshot was taken.
+
+	// For case 1, we distribute the rewards by adding them to the staker's staked amount.
+	if stakeEntry != nil && stakeEntry.RewardMethod == StakingRewardMethodRestake {
+		stakeEntry.StakeAmountNanos.Add(stakeEntry.StakeAmountNanos, uint256.NewInt().SetUint64(rewardNanos))
+		bav._setStakeEntryMappings(stakeEntry)
+
+		return nil
+	}
+
+	// For cases 2 and 3, the staker does not want their rewards restaked. The staker is still
+	// eligible to receive rewards because their stake was used to secure the network. So we pay out
+	// the rewards directly to the staker's wallet.
+
+	stakerPublicKey := bav.GetPublicKeyForPKID(stakerPKID)
+	if _, err = bav._addBalance(rewardNanos, stakerPublicKey); err != nil {
+		return errors.Wrapf(err, "distributeStakingReward: problem adding rewards to staker's DESO balance: ")
+	}
```

Comment:
Just a note that this is going to need some extra work in Rosetta to make sure its accounting tracks this. I'm not sure the right solution for this yet, and we don't need to resolve it here. Just something to keep in mind.

Likely we will need to output something to Rosetta that allows it to figure out what's going on. Or, possibly even crazier, we could rewrite Rosetta to use state syncer, which may potentially be easier and eliminate edge cases... But probably not. We'll probably just want to hack this together.

Reply:
Sounds good. I've made a not of this as a requirement for Rosetta when we eventually get to it.